File: src/main/java/leetcode/topinterview/LeetCode125.java
package leetcode.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/24 18:50
 */
public class LeetCode125 {

    public static void main(String[] args) {
        System.out.println(new LeetCode125().isPalindrome("0P"));
    }

    public boolean isPalindrome(String s) {

        if (s == null || s.length() == 0) {
            return true;
        }
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            while (!Character.isLetterOrDigit(s.charAt(left)) && left < right) {
                left++;
            }

            while (!Character.isLetterOrDigit(s.charAt(right)) && right > left) {
                right--;
            }

            //fixed
            if (left < right) {
                if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                    return false;
                }
            }
            left++;
            right--;
        }

        return true;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode125.go
package topinterview

import "strings"

func isPalindrome(s string) bool {
	left, right := 0, len(s)-1
	IsAlb := func(ch byte) bool {
		return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9')
	}
	for left < right {
		for left < right && (!IsAlb(s[left]) || s[left] == ' ') {
			left++
		}
		for left < right && (!IsAlb(s[right]) || s[right] == ' ') {
			right--
		}
		if strings.ToLower(string(s[left])) == strings.ToLower(string(s[right])) {
			left++
			right--
		} else {
			return false
		}
	}
	return true
}
